lbrynet/dht/protocol/iterative_find.py

```diff
@@ -215,70 +219,41 @@ def callback(_):
                 self.delayed_calls.append(self.loop.call_later(delay, self._search))
         except (asyncio.CancelledError, StopAsyncIteration):
             if self.running:
-                drain_tasks(self.running_probes)
-                self.running = False
+                self.loop.call_soon(self.aclose)
 
     def _search(self):
         self.tasks.append(self.loop.create_task(self._search_task()))
 
-    def search(self):
+    def __aiter__(self):
         if self.running:
             raise Exception("already running")
         self.running = True
         self._search()
-
-    async def next_queue_or_finished(self) -> typing.List['KademliaPeer']:
-        peers = self.loop.create_task(self.iteration_queue.get())
-        finished = self.loop.create_task(self.finished.wait())
-        err = None
-        try:
-            await asyncio.wait([peers, finished], loop=self.loop, return_when='FIRST_COMPLETED')
-            if peers.done():
-                return peers.result()
-            raise StopAsyncIteration()
-        except asyncio.CancelledError as error:
-            err = error
-        finally:
-            if not finished.done() and not finished.cancelled():
-                finished.cancel()
-            if not peers.done() and not peers.cancelled():
-                peers.cancel()
-            if err:
-                raise err
-
-    def __aiter__(self):
-        self.search()
         return self
 
     async def __anext__(self) -> typing.List['KademliaPeer']:
         try:
             if self.iteration_count == 0:
-                initial_results = self.get_initial_result()
-                if initial_results:
-                    self.iteration_queue.put_nowait(initial_results)
-            result = await self.next_queue_or_finished()
+                result = self.get_initial_result() or await self.iteration_queue.get()
+            else:
+                result = await self.iteration_queue.get()
+            if not result:
+                raise StopAsyncIteration
             self.iteration_count += 1
             return result
         except (asyncio.CancelledError, StopAsyncIteration):
-            await self.aclose()
+            self.loop.call_soon(self.aclose)
```

Comment:
why call_soon for this? shouldn't it block?

Reply:
because aclose can try to cancel the caller, call_soon detaches the relashionship